### PR TITLE
i#4134 drbbdup: Avoid flags preservation for 2 cases

### PR DESCRIPTION
**Pull request by derekbruening**
When there are just 2 drbbdup cases and one has an encoding of zero,
we can use a flags-free jump-if-register-is-zero for our dispatch,
avoiding flags preservation costs.

Applies this to x86 as well by switching to the xcx scratch register
and using JECXZ.  JECXZ is relatively slow on modern processors.  I
measure its performance, and it depends on the application whether it
out-performs savings the flags.  I left it as the default with hopes
that it will help more often than not on larger clients and
applications, but we can remove it if that is not borne out in future
evaluations.

The existing no-encode-test meets the criteria and serves as a test.

Before:
```
  --------------------------------------------------
  after instrumentation:
  TAG  0x0000ffff868340c0
   +0    m4 @0x0000fffd428950e8  f900b781   str    %x1 -> +0x0168(%x28)[8byte]
   +4    m4 @0x0000fffd42894da0  d53b4200   mrs    %nzcv -> %x0
   +8    m4 @0x0000fffd42894cd8  f900af80   str    %x0 -> +0x0158(%x28)[8byte]
   +12   m4 @0x0000fffd42894c58  d28c1000   movz   $0x6080 lsl $0x00 -> %x0
   +16   m4 @0x0000fffd42894bd8  f2a85000   movk   %x0 $0x4280 lsl $0x10 -> %x0
   +20   m4 @0x0000fffd42894b10  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
   +24   m4 @0x0000fffd42894a48  f9400000   ldr    (%x0)[8byte] -> %x0
   +28   m4 @0x0000fffd42894e20  f9400000   <label>
   +28   m4 @0x0000fffd42894980  f100041f   subs   %x0 $0x0000000000000001 lsl $0x0000000000000000 -> %xzr
   +32   m4 @0x0000fffd42894900  54000001   b.ne   @0x0000fffd42894fa0[8byte]
  --------------------------------------------------
```
After:
```
  --------------------------------------------------
  after instrumentation:
  TAG  0x0000ffffa53f20c0
   +0    m4 @0x0000fffd614530e8  f900b781   str    %x1 -> +0x0168(%x28)[8byte]
   +4    m4 @0x0000fffd61452da0  d28a1000   movz   $0x5080 lsl $0x00 -> %x0
   +8    m4 @0x0000fffd61452cd8  f2ac2780   movk   %x0 $0x613c lsl $0x10 -> %x0
   +12   m4 @0x0000fffd61452c58  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
   +16   m4 @0x0000fffd61452bd8  f9400000   ldr    (%x0)[8byte] -> %x0
   +20   m4 @0x0000fffd61452e20  f9400000   <label>
   +20   m4 @0x0000fffd61452b10  b4000000   cbz    @0x0000fffd61452fa0[8byte] %x0
  --------------------------------------------------
```

Issue: #4134

### CODE REVIEW

**Comment by johnfxgalea** [comment]
Would it be possible to add another variant of no-encode-test where the default case is not zero?

**Review by johnfxgalea — APPROVED** [review]

File: ext/drbbdup/drbbdup.c [hunk 1]
```diff
@@ -908,9 +970,20 @@ drbbdup_insert_dispatch(void *drcontext, instrlist_t *bb, instr_t *where,
 
     /* If runtime encoding not equal to encoding of current case, just jump to next.
      */
+    bool jmp_if_equal = false, avoid_flags = false;
+    if (drbbdup_case_zero_vs_nonzero(manager)) {
+        /* Use an aflags-less jump-if-zero. */
+        avoid_flags = true;
+        if (current_case->encoding != 0) {
+            /* Invert the compare. */
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply ext/drbbdup/drbbdup.c:1009]: Done.

**Comment by derekbruening** [comment]
> Would it be possible to add another variant of no-encode-test where zero is not the default case but another case?

Yes, done.
